--- r5p0/src/devicedrv/mali/linux/mali_kernel_linux.c.orig	2014-11-11 07:57:22.000000000 +0100
+++ r5p0/src/devicedrv/mali/linux/mali_kernel_linux.c	2018-12-02 20:36:49.000000000 +0100
@@ -364,7 +364,7 @@
 	int err = 0;
 
 	MALI_DEBUG_PRINT(2, ("Inserting Mali v%d device driver. \n", _MALI_API_VERSION));
-	MALI_DEBUG_PRINT(2, ("Compiled: %s, time: %s.\n", __DATE__, __TIME__));
+	//MALI_DEBUG_PRINT(2, ("Compiled: %s, time: %s.\n", __DATE__, __TIME__));
 	MALI_DEBUG_PRINT(2, ("Driver revision: %s\n", SVN_REV_STRING));
 
 #if MALI_ENABLE_CPU_CYCLES
